test/grad_test.jl

```diff
@@ -0,0 +1,71 @@
+using Test
+using Printf
+using LinearAlgebra
+mean(x) = sum(x)/length(x)
+
+function log_division(a, b)
+    if b == 0
+        return a == 0 ? NaN : Inf
+    end
+    return log(a / b)
+end
+
+function grad_test(f, x0, dx, dJdx; dJ=nothing, maxiter=6, h0=5e-2, stol=1e-1, hfactor=8e-1)
```

Comment:
do you have a reference for this test? It looks nice

Reply:
I'll add some document to it, but it's essentially running the same test with some more printouts that I have found useful when I had to debug gradient calculations with Pyadjoint. Nearly all of my experience with gradient computations come from [Pyadjoint](https://www.dolfin-adjoint.org/en/latest/documentation/verification.html) and my usage of it in [CRIKit](https://gitlab.com/crikit/crikit). Pyadjoint has a good set of [mathematical documentation](https://www.dolfin-adjoint.org/en/latest/documentation/maths/index.html), .